### PR TITLE
[vtctld] Add v0 GetWorkflows rpc and workflow/vexec packages

### PR DESCRIPTION
**Pull request by ajm188**
## Description

This adds an initial attempt at a GetWorkflows rpc to VtctldServer, by way of introducing new `vtctl/workflow` and `vtctl/workflow/vexec` packages to eventually share more code between the legacy and new vtctl(d) servers. I also made a more committed effort on that front by reimplementing wrangler's ListWorkflows function using the new vexec.

### `package go/vt/vtctl/workflow`

This provides the actual implementation of GetWorkflows. I'm debating of separating the GetWorkflows rpc into a separate "service" in the protobuf file, because it _seems_ weird to me that `workflow.Server` provides the actual implementation. I think this would look something like:

```protobuf
service Workflow {
  rpc GetWorkflows(vtctldata.GetWorkflowsRequest) returns (vtctldata.GetWorkflowsResponse) {};
}
```

```go
// vtctl/grpcvtctldserver/server.go

func StartServer(s *grpc.Server, ts *topo.Server) {
	vtctld := NewVtctldServer(ts)
	vtctlservicepb.RegisterWorkflowServer(s, vtctld)
	vtctlservicepb.RegisterVtctldServer(s, vtctld)
}

// vtctl/workflow/server.go

type Server struct { ... }

var _ vtctlservicepb.WorkflowServer = (*Server)(nil)
```

This would allow us to assert that `workflow.Server` correctly implements the entirety of the Workflow rpcs, without also needing to implement the rest of VtctldServer by separating out the concerns. But then registering is maybe weird, so idk. All perspectives welcome!!

### `package go/vt/vtctl/workflow/vexec`

This attempts to create a more ... encapsulated vexec API by more cleanly separating the query parsing/planning/rewriting phase from the execution phase. Eventually this should include all the functionality of wrangler's `runVexec`, but not yet, because that was a lot.

### Usage

I got a Reshard running by running up through 301_reshard.sh of the local example, and then:

<details><summary>New CLI / code</summary>

```
❯ vtctldclient --server "localhost:15999" GetWorkflows customer
{
  "workflows": [
    {
      "name": "cust2cust",
      "source": {
        "keyspace": "customer",
        "shards": [
          "0"
        ]
      },
      "target": {
        "keyspace": "customer",
        "shards": [
          "-80",
          "80-"
        ]
      },
      "max_v_replication_lag": "1",
      "shard_statuses": {
        "-80/zone1-0000000300": {
          "primary_replication_statuses": [
            {
              "id": "1",
              "shard": "-80",
              "tablet": {
                "cell": "zone1",
                "uid": 300
              },
              "binlog_source": {
                "keyspace": "customer",
                "shard": "0",
                "tablet_type": "UNKNOWN",
                "key_range": null,
                "tables": [
                ],
                "filter": {
                  "rules": [
                    {
                      "match": "/.*",
                      "filter": "-80"
                    }
                  ],
                  "fieldEventMode": "ERR_ON_MISMATCH"
                },
                "on_ddl": "IGNORE",
                "external_mysql": "",
                "stop_after_copy": false
              },
              "position": "MySQL56/715d1016-7ab7-11eb-b361-248943b3b987:1-38",
              "stop_position": "",
              "state": "Running",
              "db_name": "vt_customer",
              "transaction_timestamp": {
                "seconds": "0",
                "nanoseconds": 0
              },
              "time_updated": {
                "seconds": "1614622642",
                "nanoseconds": 0
              },
              "message": "",
              "copy_states": [
              ]
            }
          ],
          "tablet_controls": [
          ],
          "is_primary_serving": false
        },
        "80-/zone1-0000000400": {
          "primary_replication_statuses": [
            {
              "id": "1",
              "shard": "80-",
              "tablet": {
                "cell": "zone1",
                "uid": 400
              },
              "binlog_source": {
                "keyspace": "customer",
                "shard": "0",
                "tablet_type": "UNKNOWN",
                "key_range": null,
                "tables": [
                ],
                "filter": {
                  "rules": [
                    {
                      "match": "/.*",
                      "filter": "80-"
                    }
                  ],
                  "fieldEventMode": "ERR_ON_MISMATCH"
                },
                "on_ddl": "IGNORE",
                "external_mysql": "",
                "stop_after_copy": false
              },
              "position": "MySQL56/715d1016-7ab7-11eb-b361-248943b3b987:1-38",
              "stop_position": "",
              "state": "Running",
              "db_name": "vt_customer",
              "transaction_timestamp": {
                "seconds": "0",
                "nanoseconds": 0
              },
              "time_updated": {
                "seconds": "1614622642",
                "nanoseconds": 0
              },
              "message": "",
              "copy_states": [
              ]
            }
          ],
          "tablet_controls": [
          ],
          "is_primary_serving": false
        }
      }
    }
  ]
}
```

</details>

<details><summary>Old CLI / code</summary>

```
❯ vtctlclient -server "localhost:15999" Workflow customer.cust2cust show
{
        "Workflow": "cust2cust",
        "SourceLocation": {
                "Keyspace": "customer",
                "Shards": [
                        "0"
                ]
        },
        "TargetLocation": {
                "Keyspace": "customer",
                "Shards": [
                        "-80",
                        "80-"
                ]
        },
        "MaxVReplicationLag": 0,
        "ShardStatuses": {
                "-80/zone1-0000000300": {
                        "MasterReplicationStatuses": [
                                {
                                        "Shard": "-80",
                                        "Tablet": "zone1-0000000300",
                                        "ID": 1,
                                        "Bls": {
                                                "keyspace": "customer",
                                                "shard": "0",
                                                "filter": {
                                                        "rules": [
                                                                {
                                                                        "match": "/.*",
                                                                        "filter": "-80"
                                                                }
                                                        ]
                                                }
                                        },
                                        "Pos": "MySQL56/715d1016-7ab7-11eb-b361-248943b3b987:1-38",
                                        "StopPos": "",
                                        "State": "Running",
                                        "DBName": "vt_customer",
                                        "TransactionTimestamp": 0,
                                        "TimeUpdated": 1614631626,
                                        "Message": "",
                                        "CopyState": null
                                }
                        ],
                        "TabletControls": null,
                        "MasterIsServing": false
                },
                "80-/zone1-0000000400": {
                        "MasterReplicationStatuses": [
                                {
                                        "Shard": "80-",
                                        "Tablet": "zone1-0000000400",
                                        "ID": 1,
                                        "Bls": {
                                                "keyspace": "customer",
                                                "shard": "0",
                                                "filter": {
                                                        "rules": [
                                                                {
                                                                        "match": "/.*",
                                                                        "filter": "80-"
                                                                }
                                                        ]
                                                }
                                        },
                                        "Pos": "MySQL56/715d1016-7ab7-11eb-b361-248943b3b987:1-38",
                                        "StopPos": "",
                                        "State": "Running",
                                        "DBName": "vt_customer",
                                        "TransactionTimestamp": 0,
                                        "TimeUpdated": 1614631626,
                                        "Message": "",
                                        "CopyState": null
                                }
                        ],
                        "TabletControls": null,
                        "MasterIsServing": false
                }
        }
}
```

</details>

## Related Issue(s)

- #7574

## Checklist
- [ ] Should this PR be backported? **No**
- [x] Tests were added or are not required
  - I definitely deferred testing some things to try to get this reviewed. I'll definitely come back for these!!
- [x] Documentation was added or is not required

## Follow-up Tasks
- [ ] File issues to add additional tests
- [ ] File issues to port more functionality from wrangler vexec to this vexec, and backwards-reimplement existing APIs

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [x]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

File: proto/vtctldata.proto [hunk 1]
```diff
@@ -86,14 +86,20 @@ message Workflow {
...
   }
 
-  message ReplicationStatus {
+  message ShardStream {
+    repeated Stream streams = 1;
+    repeated topodata.Shard.TabletControl tablet_controls = 2;
+    bool is_primary_serving = 3;
+  }
+
+  message Stream {
```
Comment (ajm188) [inline proto/vtctldata.proto:102]: @rohit-nayak-ps let me know what you think of the new naming

File: go/vt/vtctl/workflow/vexec/query_planner.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+// VReplicationQueryPlanner is a good example implementation to refer to.
+type QueryPlanner interface {
+	// (NOTE:@ajm188) I don't think this method fits on the query planner. To
+	// me, especially given that it's only implemented by the vrep query planner
+	// in the old implementation (the schema migration query planner no-ops this
+	// method), this fits better on our workflow.Manager struct, probably as a
+	// method called something like "VReplicationExec(ctx, query, Options{DryRun: true})"
+	// DryRun(ctx context.Context) error
+
+	// PlanQuery contsructs and returns a QueryPlan for a given statement. The
```
Comment (rohit-nayak-ps) [inline go/vt/vtctl/workflow/vexec/query_planner.go:63]: typo: contsructs => constructs

**Review by rohit-nayak-ps — APPROVED** [review]
very nice

File: proto/vtctldata.proto (hunk 1 above)
Comment (rohit-nayak-ps) [reply proto/vtctldata.proto:106]: This is good!

File: go/vt/vtctl/workflow/doc.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	existing wrangler code to adapt to it, than make a subpar API in the name of
+	backwards compatibility. I'm not sure if that's a tradeoff I'll even need to
+	consider in the future, but I'm putting a stake in the ground on which side
+	of that tradeoff I intend to fall, should it come to it.
+(3) Eventually we'll need to consider how the online schema migration workflows
+	fit into this. I'm trying to at least be somewhat abstract in the
+	vexec / queryplanner APIs to fit with the QueryParams thing that wrangler
+	uses, which _should_ work, but who knows?? Time will tell.
+*/
+package workflow
```
[the remark this answers is not in the document]
Comment (ajm188) [reply go/vt/vtctl/workflow/doc.go:45]: Yeah, this is definitely not as clear as could be; I'd prefer you didn't have to read all the code to understand what this is trying to say 😅 

I can definitely try to clean this up in a later PR (probably when I add the `GetWorkflow`), as I want to get this merged soon-ish to unblock us in vtadmin-land to start working with this API.

**Pull request merged**